### PR TITLE
[grafana] Add support for extra volumes for Grafana Image Renderer

### PR DESCRIPTION
**Pull request by bradleypettit**
Adds support for specificying extraConfigmapMounts, extraSecretMounts, and extraVolumeMounts for the image renderer pods.

This is required if you need to do more advanced configuration of the image renderer via a JSON configuration file. This should unblock issues like https://github.com/grafana/helm-charts/issues/2310.

This will also allow people to configure trusted CA certs for the image renderer pods, resolving issues like https://github.com/grafana/helm-charts/issues/1267.

### CODE REVIEW

**Review by zalegrala — APPROVED** [review]

File: charts/grafana/templates/image-renderer-deployment.yaml [hunk 1]
```diff
@@ -109,6 +109,24 @@ spec:
           volumeMounts:
             - mountPath: /tmp
               name: image-renderer-tmpfs
+            {{- range .Values.imageRenderer.extraConfigmapMounts }}
+            - name: {{ tpl .name $root }}
+              mountPath: {{ tpl .mountPath $root }}
```
Comment (jkroepke) [inline charts/grafana/templates/image-renderer-deployment.yaml:114]: Could you please explain, why tpl is required here?

Comment (bradleypettit) [reply charts/grafana/templates/image-renderer-deployment.yaml:114]: In this case it's just to remain consistent with the same options in the Grafana deployment. I'm not sure why templating was only enabled for extraConfigmapMounts and not extraSecretMounts/extraVolumeMounts, but it was introduced in this PR - https://github.com/grafana/helm-charts/pull/1333.

Comment (jkroepke) [reply charts/grafana/templates/image-renderer-deployment.yaml:114]: Thanks for clarification!

File: charts/grafana/values.yaml [hunk 2]
```diff
@@ -1248,6 +1248,18 @@ imageRenderer:
...
+  extraConfigmapMounts: []
+
+  # Extra secrets to mount in image-renderer pods
+  extraSecretMounts: []
+
+  # Extra volumes to mount in image-renderer pods
+  extraVolumeMounts: []
+
+  # Extra volumes for image-renderer pods
+  extraVolumes: []
```
Comment (jkroepke) [inline charts/grafana/values.yaml:1261]: I miss one more thing:

Could you please fillout some new extra values in the CI test config that this values are covered by CI?

https://github.com/grafana/helm-charts/blob/main/charts/grafana/ci/with-image-renderer-values.yaml

For extraVolumes / extraVolumeMounts, empty would be sufficient. 

For extraConfigmapMounts / extraSecretMounts, you may have to setup own ConfigMaps / Secrets, which can be done via extraObject

https://github.com/bradleypettit/helm-charts-grafana-renderer-support-volumes/blob/16c6ee86bb95d69488e0e9cdfd4cd27cbbf30b50/charts/grafana/values.yaml#L1332

File: charts/grafana/templates/image-renderer-deployment.yaml [hunk 3]
```diff
@@ -128,4 +146,54 @@ spec:
...
+            {{ toYaml .hostPath | nindent 6 }}
+          {{- else if .csi }}
+          csi:
+            {{- toYaml .csi | nindent 6 }}
+          {{- else if .configMap }}
+          configMap:
+            {{- toYaml .configMap | nindent 6 }}
+          {{- else if .emptyDir }}
+          emptyDir:
+            {{- toYaml .emptyDir | nindent 6 }}
```
Comment (jkroepke) [inline charts/grafana/templates/image-renderer-deployment.yaml:194]: I count 12 indent here.

File: charts/grafana/templates/image-renderer-deployment.yaml [hunk 4]
```diff
@@ -128,4 +146,54 @@ spec:
...
+            claimName: {{ .existingClaim }}
+          {{- else if .hostPath }}
+          hostPath:
+            {{ toYaml .hostPath | nindent 6 }}
+          {{- else if .csi }}
+          csi:
+            {{- toYaml .csi | nindent 6 }}
+          {{- else if .configMap }}
+          configMap:
+            {{- toYaml .configMap | nindent 6 }}
```
Comment (jkroepke) [inline charts/grafana/templates/image-renderer-deployment.yaml:191]: I count 12 indent here.

File: charts/grafana/templates/image-renderer-deployment.yaml [hunk 5]
```diff
@@ -128,4 +146,54 @@ spec:
...
+        - name: {{ .name }}
+          {{- if .existingClaim }}
+          persistentVolumeClaim:
+            claimName: {{ .existingClaim }}
+          {{- else if .hostPath }}
+          hostPath:
+            {{ toYaml .hostPath | nindent 6 }}
+          {{- else if .csi }}
+          csi:
+            {{- toYaml .csi | nindent 6 }}
```
Comment (jkroepke) [inline charts/grafana/templates/image-renderer-deployment.yaml:188]: I count 12 indent here.

File: charts/grafana/templates/image-renderer-deployment.yaml [hunk 6]
```diff
@@ -128,4 +146,54 @@ spec:
...
+        {{- end }}
+        {{- end }}
+        {{- range .Values.imageRenderer.extraVolumes }}
+        - name: {{ .name }}
+          {{- if .existingClaim }}
+          persistentVolumeClaim:
+            claimName: {{ .existingClaim }}
+          {{- else if .hostPath }}
+          hostPath:
+            {{ toYaml .hostPath | nindent 6 }}
```
Comment (jkroepke) [inline charts/grafana/templates/image-renderer-deployment.yaml:185]: I count 12 indent here.

File: charts/grafana/templates/image-renderer-deployment.yaml [hunk 7]
```diff
@@ -128,4 +146,54 @@ spec:
...
+              {{- toYaml . | nindent 8 }}
+            {{- end }}
+        {{- else if .projected }}
+        - name: {{ .name }}
+          projected:
+            {{- toYaml .projected | nindent 6 }}
+        {{- else if .csi }}
+        - name: {{ .name }}
+          csi:
+            {{- toYaml .csi | nindent 6 }}
```
Comment (jkroepke) [inline charts/grafana/templates/image-renderer-deployment.yaml:175]: I count 12 indent here.

File: charts/grafana/templates/image-renderer-deployment.yaml [hunk 8]
```diff
@@ -128,4 +146,54 @@ spec:
...
+            secretName: {{ .secretName }}
+            defaultMode: {{ .defaultMode }}
+            {{- with .items }}
+            items:
+              {{- toYaml . | nindent 8 }}
+            {{- end }}
+        {{- else if .projected }}
+        - name: {{ .name }}
+          projected:
+            {{- toYaml .projected | nindent 6 }}
```
Comment (jkroepke) [inline charts/grafana/templates/image-renderer-deployment.yaml:171]: I count 12 indent here.

File: charts/grafana/templates/image-renderer-deployment.yaml [hunk 9]
```diff
@@ -128,4 +146,54 @@ spec:
...
+        {{- end }}
+        {{- range .Values.imageRenderer.extraSecretMounts }}
+        {{- if .secretName }}
+        - name: {{ .name }}
+          secret:
+            secretName: {{ .secretName }}
+            defaultMode: {{ .defaultMode }}
+            {{- with .items }}
+            items:
+              {{- toYaml . | nindent 8 }}
```
Comment (jkroepke) [inline charts/grafana/templates/image-renderer-deployment.yaml:166]: please check this indent

File: charts/grafana/templates/image-renderer-deployment.yaml [hunk 10]
```diff
@@ -128,4 +146,54 @@ spec:
       volumes:
         - name: image-renderer-tmpfs
           emptyDir: {}
+        {{- range .Values.imageRenderer.extraConfigmapMounts }}
+        - name: {{ tpl .name $root }}
+          configMap:
+            name: {{ tpl .configMap $root }}
+            {{- with .items }}
+            items:
+              {{- toYaml . | nindent 8 }}
```
Comment (jkroepke) [inline charts/grafana/templates/image-renderer-deployment.yaml:155]: please check this indent

**Review by jkroepke — CHANGES_REQUESTED** [review]
I found some suspicious indents. Maybe they work on grafana pod, since the grafana pod is inside a template.

**Comment by bradleypettit** [comment]
Thanks @jkroepke - corrected the indenting/linting and added to the CI test.

**Review by jkroepke — APPROVED** [review]
Great! Thanks!

**Pull request merged**